### ISSUE TITLE
Mismatching time-step (dt) and total-time values after the decimal point

### ISSUE THREAD

**Issue by cy16f01** [MDAnalysis/mdanalysis#1886]
I have performed simulation in GROMACS, for dt=0.001ps, total frames =1000001, total time=1000ps(1ns).

But in MDAnalysis if i give the commands as:-
```python
import MDAnalysis as mda
u=mda.Universe('nptmd1f.tpr','nptmd1f.xtc')
print(u.trajectory.ts.dt)
print(u.trajectory.totaltime)
print(u.trajectory.n_frames)
print(len(u.trajectory)) 
```

I am getting the output as:-
0.0010000000474974513
1000.0000474974513
1000001
1000001
But actually it should be (I am expecting) as:-
0.001
1000.00
1000001
1000001

How are the values after decimal place getting varied..? It should not vary right..?? This values after the decimal values ie., say 0.001000000047494513 (which should be 0.001) does affect the HB auto-correlation code timings (dt/timestep etc...) and hence affect and leads to wrong values...
Is this a bug..??
Any suggestions are appreciated.

**Comment by jbarnoud** [comment]
This looks like rounding errors due to how floating numbers are stored in memory. The error is very small compared to the actual value, I do not expect it to have a noticeable influence on anything.

**Comment by cy16f01** [comment]
Yes, this is rounding errors. And each of these very small errors can add up and hence lead to significant error value.
This lead to the different time interval values in the results say for example when i am running the HB auto-correlation code, then (since the hydrogen bonds making and breaking are in few femtoseconds..) i got the following mismatch in time values...
In the following code i have run :- 
`python
import MDAnalysis as mda
from MDAnalysis.analysis import hbonds
import matplotlib.pyplot as plt
%matplotlib inline
u = mda.Universe('nptmd1f.tpr','nptmd1f.xtc')
H = u.select_atoms('byres(name OW HW1 HW2 and around 5.6 name CA)')
A = u.select_atoms('byres(name OW HW1 HW2 and around 5.6 name CA)')
D = u.select_atoms('byres(name OW HW1 HW2 and around 5.6 name CA)')
hb_ac = hbonds.HydrogenBondAutoCorrel(u, acceptors=A, hydrogens=H, donors=D,bond_type='continuous',
                            sample_time=10,nruns=20000,nsamples=20,pbc=True,angle_crit=130.0, dist_crit=3.0)
hb_ac.run()
hb_ac.solve()
time = hb_ac.solution['time']
results = hb_ac.solution['results']
tau = hb_ac.solution['tau']
fit=hb_ac.solution['fit']
estimate = hb_ac.solution['estimate']
print("{time} {results}".format(time=time,results=results))
print("{time} {estimate}".format(time=time,estimate=estimate))
plt.figure(1,figsize=(15, 5))
plt.figure(1,figsize=(15, 5))
plt.subplot(122)
plt.xlabel('time')
plt.ylabel('HBLc')
plt.title('HBL Continuos')
plt.plot(time,results, 'ro')
plt.plot(time,estimate)
plt.show()
print (tau)
`

**Comment by richardjgowers** [comment]
So the time is recorded as `self.solution['time'] = np.arange(len(master_results), dtype=np.float32) * self.u.trajectory.dt * self._skip`, so the slight rounding error is amplified for later times.  If the trajectory provides time, it should probably get used to avoid rounding errors like this.  I think it's probably not going to affect the overall results by much though.

**Comment by cy16f01** [comment]
But in the previous simulation, when i had saved trajectory   dt=0.5ps, total frames =20000, total time=10000ps(10ns) and i gave the same code i got the decimal values exact.
```python
import MDAnalysis as mda
u=mda.Universe('nptmd0.5.tpr','nptmd0.5.xtc')
print(u.trajectory.ts.dt)
print(u.trajectory.totaltime)
print(u.trajectory.n_frames)
print(len(u.trajectory)) 
```
0.05
10000.0
20001
20001
Now when i reduced the dt from 0.5ps to 0.001ps (to save more trajectories and hence to calculate HB), I am getting the rounding off errors. Why is this happenning when i reduce the dt ..?? Actually it should show the exact values right..??

**Comment by richardjgowers** [comment]
@cy16f01 https://en.wikipedia.org/wiki/Floating-point_arithmetic#Representable_numbers,_conversion_and_rounding

It's just how numbers are stored in computers.

**Comment by orbeckst** [comment]
Should we store times as float64? It won't solve all problems (and just postpone the issue) but it might help in simple cases. 

As https://github.com/richardjgowers said: dealing with floating point numbers is difficult. If you know that you need higher accuracy then you have to think carefully about designing algorithms. Perhaps in this case you could get away with using integer time steps for the calculation and then multiply with the dt in the end.
